os/drivers/pm/pm.c

```diff
@@ -0,0 +1,151 @@
+
+/****************************************************************************
+ * Function Prototypes
+ ****************************************************************************/
+
+static int pm_ioctl(FAR struct file *filep, int cmd, unsigned long arg);
+static ssize_t pm_read(FAR struct file *filep, FAR char *buffer, size_t len);
+static ssize_t pm_write(FAR struct file *filep, FAR const char *buffer, size_t len);
+
+/****************************************************************************
+ * Data
+ ****************************************************************************/
+static const struct file_operations pm_fops = {
+	0,                          /* open */
+	0,                          /* close */
+	pm_read,                    /* read */
+	pm_write,                   /* write */
+	0,                          /* seek */
+	pm_ioctl                    /* ioctl */
+#ifndef CONFIG_DISABLE_POLL
+	, 0                         /* poll */
+#endif
+};
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+static ssize_t pm_read(FAR struct file *filep, FAR char *buffer, size_t len)
+{
+	return 0;
+}
+
+static ssize_t pm_write(FAR struct file *filep, FAR const char *buffer, size_t len)
+{
+	return 0;
+}
+
+
+/************************************************************************************
+ * Name: pm_ioctl
+ *
+ * Command description:
+ *   PMIOC_LOCK - for locking PM transition 
+ *   PMIOC_UNLOCK - for unlocking PM transiton
+ *   PMIOC_TIMER_LOCK - for locking PM transition for certain time interval
+ *   PMIOC_TIMER_SET - to set a wakeup timer
+ *   PMIOC_TIMER_CANCEL - to stop a wakeup timer
+ *   PMIOC_TUNEFREQ - for changing the operating frequency of the core to save power
+ * 
+ * Arguments:
+ *   filep is ioctl fd, cmd is required command, arg is required argument for
+ *   the command. 
+ *   for TIMER_LOCK, arg should be a long unsigned type (timer interval in micro second).
+ *   for TIMER_SET, arg should be an unsigned int.
+ *   for TIMER_CANCEL, arg not required.
+ *   for TUNEFREQ, arg should be an int type.
+ *
+ * Description:
+ *   This api can be used to perform PM operation.
+ *
+ ************************************************************************************/
+static int pm_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+	int ret = -ENOSYS;
+    
+	/* Handle built-in ioctl commands */
+	switch (cmd) {
+    	case PMIOC_LOCK:
+        	pm_stay(PM_IDLE_DOMAIN, PM_NORMAL);
+        	pmvdbg("State locked!\n");
+        	ret = OK;
+        	break;
+        case PMIOC_UNLOCK:
+            	pm_relax(PM_IDLE_DOMAIN, PM_NORMAL);
+            	pmvdbg("State unlocked!\n");
+            	ret = OK;
+            	break;
+	case PMIOC_TIMER_LOCK:
+		if (arg > 0) {
+			pm_stay(PM_IDLE_DOMAIN, PM_NORMAL);
+			pmvdbg("State locked!\n");
+			pm_set_timer(PM_LOCK_TIMER, arg);
```

Comment:
pm_set_timer is h/w timer.

we don't need h/w timer in pm timer of locking.
Please use sw timer(wdog)